llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -3914,8 +3914,12 @@ LegalizerHelper::lower(MachineInstr &MI, unsigned TypeIdx, LLT LowerHintTy) {
   case G_SSHLSAT:
   case G_USHLSAT:
     return lowerShlSat(MI);
-  case G_ABS:
+  case G_ABS: {
+    LLT Ty = MRI.getType(MI.getOperand(0).getReg());
+    if (LI.isLegalOrCustom({G_ABS, Ty}))
```

Comment:
Checking the G_ABS legality doesn't make sense in the G_ABS lowering implementation. You should need to check the operations in the expansion

Reply:
I think the problem might be that there is no good way with the isLegal calls to check if there is a single instruction that does "select and conditional neg" at the same time, in the way csneg would do.

Reply:
For reference, the DAG has this chain of implementations: https://github.com/llvm/llvm-project/blob/8f68022f8e6e54d1aeae4ed301f5a015963089b7/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L8962

Reply:
Checking for G_SELECT and G_ICMP leads to assertion when compiled for  AMDGPU in abs.ll.
`const T &llvm::ArrayRef<llvm::LLT>::operator[](size_t) const [T = llvm::LLT]: Assertion `Index < Length && "Invalid index!"' failed.`

Reply:
```
LLT Ty = MRI.getType(MI.getOperand(0).getReg());
dbgs() << LI.isLegalOrCustom({G_ICMP, Ty});
```
Leads to an assert for llvm.abs.ll test. This doesn't need my patch to trigger an assert. 
@arsenm Thoughts? 
If this is uncovering a latent issue then I would like to go with legalizeCustom() and follow @davemgreen's suggestion. This also aligns with what RISCV is doing.

Reply:
Checking for them how? 

The specific instruction AArch64 wants to use is only indirectly related

Reply:
Found that the assert is triggering for `llvm.abs.i64` case in `AMDGPU/GlobalISel/llvm.abs.ll`. Constraining the lowering to < 8 bytes has no assert. Do you know what might be causing this?

Reply:
Don't know what that has to do with abs, but this isn't a well formed query for G_ICMP legality. G_ICMP has 2 type indexes

Reply:
Hmm... I was looking at how G_SSUBSAT is using G_SMIN and followed the same approach. 
Anyway, updated the branch to check both type indexes.